core/src/main/java/com/linecorp/armeria/internal/common/TlsProviderUtil.java

```diff
@@ -0,0 +1,209 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import static java.util.Objects.requireNonNull;
+
+import java.net.IDN;
+import java.time.Duration;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.github.benmanes.caffeine.cache.Scheduler;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.TlsKeyPair;
+import com.linecorp.armeria.common.TlsProvider;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.metric.CloseableMeterBinder;
+import com.linecorp.armeria.common.metric.MeterIdPrefix;
+import com.linecorp.armeria.common.metric.MoreMeterBinders;
+import com.linecorp.armeria.common.util.ThreadFactories;
+import com.linecorp.armeria.common.util.TlsEngineType;
+import com.linecorp.armeria.internal.common.util.SslContextUtil;
+
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.util.ReferenceCountUtil;
+
+public final class TlsProviderUtil {
+
+    @VisibleForTesting
+    static final LoadingCache<TlsCacheContext, SslContext> sslContextCache;
+
+    static {
+        final ScheduledExecutorService cacheExecutor = Executors.newSingleThreadScheduledExecutor(
+                ThreadFactories.newThreadFactory("armeria-tls-cache-executor", true));
+
+        sslContextCache =
+                Caffeine.newBuilder()
+                        .maximumSize(1024)
+                        .scheduler(Scheduler.forScheduledExecutorService(cacheExecutor))
+                        // Remove the cached SslContext after 1 hour of inactivity.
+                        .expireAfterAccess(Duration.ofHours(1))
+                        .removalListener((TlsCacheContext key, SslContext value, RemovalCause cause) -> {
+                            if (value != null) {
+                                ReferenceCountUtil.release(value);
+                            }
+                            if (key != null) {
+                                // Remove the registered metrics.
+                                key.meterBinder.close();
+                            }
+                        })
+                        .build(TlsProviderUtil::newSslContext);
+    }
+
+    public static SslContext getOrCreateSslContext(TlsProvider tlsProvider, @Nullable TlsKeyPair tlsKeyPair,
+                                                   SslContextType type, TlsEngineType tlsEngineType) {
+        final TlsCacheContext key = new TlsCacheContext(tlsProvider, tlsKeyPair, type, tlsEngineType);
```

Comment:
Question) Why does it matter that `tlsProvider` is also included in the cache key? I'm asking because I'm imagining scenarios where the `WebClient` provides the `tlsKeyPair` directly.

e.g.
```suggestion
        final TlsCacheContext key = new TlsCacheContext(tlsKeyPair, type, tlsEngineType);
```

Reply:
I thought of a similar idea before. Since `TlsKeyPair` is used only when creating a connection, it was difficult/ambiguous to utilize `TlsKeyPair` of `WebClient` which operates at the request layer. In the end, the code related to `WebClient` was deleted and I made `TlsProvider` set only through `ClientFactoryBuilder`.

Reply:
> Why does it matter that tlsProvider is also included in the cache key?
> it was difficult/ambiguous to utilize TlsKeyPair of WebClient which operates at the request layer.

I wasn't really asking about `WebClient`.

If a `tlsProvider` changes its parameters dynamically, (e.g. `allowsUnsafeCiphers` is changed from `true` to `false`), will this be supported with the current caching mechanism?

Is the expectation users update the `hashCode`, `equals` methods accordingly when dynamic parameters are modified?